xla/backends/gpu/codegen/custom.cc

```diff
@@ -1100,33 +1119,47 @@ absl::StatusOr<FusionEmissionResult> EmitCollective(
   Thunk::ThunkInfo thunk_info = Thunk::ThunkInfo::WithProfileAnnotation(instr);
 
   FusionEmissionResult result;
-  std::vector<std::unique_ptr<BufferAllocation>> fake_allocations(2);
+  std::vector<std::unique_ptr<BufferAllocation>> fake_allocations(num_args);
   if (isDynamic) {
     // Provide fake allocations for inputs and outputs. The dynamic-slice thunk
     // will own these allocations.
     unsigned fake_arg_idx = 0;
-    int64_t operand_byte_size =
-        ShapeUtil::ByteSizeOf(instr->operand(fake_arg_idx)->shape());
-    fake_allocations[fake_arg_idx] = std::make_unique<BufferAllocation>(
-        /*index=*/fake_arg_idx, operand_byte_size, /*color=*/0);
-    src = BufferAllocation::Slice(
-        /*allocation=*/fake_allocations[fake_arg_idx].get(), /*offset=*/0,
-        /*size=*/operand_byte_size);
-    fake_arg_idx++;
-    TF_RET_CHECK(instr->shape().IsArray() &&
-                 "The output is not expected to be a tuple.");
-    int64_t out_fake_byte_size =
-        ShapeUtil::ByteSizeOf(instr->shape());  // TODO: we don't need this
-    fake_allocations[fake_arg_idx] = std::make_unique<BufferAllocation>(
-        /*index=*/fake_arg_idx, /*size*/ out_fake_byte_size, /*color=*/0);
-    dst = BufferAllocation::Slice(
-        /*allocation=*/fake_allocations[fake_arg_idx].get(),
-        /*offset=*/0, /*size=*/out_fake_byte_size);
+    for (HloInstruction* operand : instr->operands()) {
+      int64_t operand_byte_size = ShapeUtil::ByteSizeOf(operand->shape());
+      fake_allocations[fake_arg_idx] = std::make_unique<BufferAllocation>(
+          /*index=*/fake_arg_idx, operand_byte_size, /*color=*/0);
+      BufferAllocation::Slice fake_slice(
+          /*allocation=*/fake_allocations[fake_arg_idx].get(), /*offset=*/0,
+          /*size=*/operand_byte_size);
+      fake_arguments.push_back(fake_slice);
+      fake_arg_idx++;
+    }
+    PtrVec<const HloInstruction*> users;
+    if (instr->shape().IsTuple()) {
+      for (const HloInstruction* user : instr->users()) {
+        users.push_back(user);
+      }
+    } else {
+      users.clear();
+      users.push_back(instr);
+    }
+    for (const HloInstruction* user : users) {
+      int64_t out_fake_byte_size =
+          ShapeUtil::ByteSizeOf(user->shape());  // TODO: we don't need this
```

Comment:
I dont understand this TODO. Can you make the comment clearer?

Reply:
This was in the other MR. Rebased with that one.